python/ray/data/tests/test_stats.py

```diff
@@ -133,6 +135,81 @@ def test_dataset_stats_sort(ray_start_regular_shared):
     assert "sort_reduce" in stats, stats
 
 
+@patch("ray.data._internal.plan.logger")
+def test_dataset_stats_autolog(mock_logger, ray_start_regular_shared):
+    """Tests functionality to autolog stats after each stage execution,
+    enabled via the `DatasetContext.enable_auto_log_stats` parameter."""
+    context = DatasetContext.get_current()
+    enable_auto_log_stats_orig = context.enable_auto_log_stats
+
+    try:
+        context.enable_auto_log_stats = True
+        ds = ray.data.range(1000, parallelism=10)
+        # Check logged stats after each stage
+        ds = ds.map_batches(lambda x: x)
+        logger_args, logger_kwargs = mock_logger.info.call_args
+        assert (
+            canonicalize(logger_args[0])
+            == """Stage N read->map_batches: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+"""
+        )
+
+        ds = ds.map(lambda x: x)
+        logger_args, logger_kwargs = mock_logger.info.call_args
+        assert (
+            canonicalize(logger_args[0])
+            == """Stage N map: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+"""
+        )
+
+        # Ensure that manual `stats()` method returns summarized stats
+        for batch in ds.iter_batches():
+            pass
+
+        stats = canonicalize(ds.stats())
+        assert (
+            stats
+            == """Stage N read->map_batches: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+
+Stage N map: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+
+Dataset iterator time breakdown:
+* In ray.wait(): T
+* In ray.get(): T
+* In next_batch(): T
+* In format_batch(): T
+* In user code: T
+* Total time: T
+"""
+        )
+    finally:
+        context.enable_auto_log_stats = enable_auto_log_stats_orig
```

Comment:
Nit: There's a pytest fixture pattern that we've used for this "enable config, run test, always make sure the config is restored" pattern that would remove the need for the try-finally block. E.g., see the [`enable_automatic_tensor_extension_cast` fixture](https://github.com/ray-project/ray/blob/2d0d6d1422623a221fb86624b583840a4a96312e/python/ray/data/tests/conftest.py#L261-L267) and a [corresponding test](https://github.com/ray-project/ray/blob/21f059ba03209840dfde478e4ab8299cbde34194/python/ray/data/tests/test_dataset.py#L929-L945) for usage. Could we do something similar here for these tests?

Reply:
got it, thanks for the detailed examples. i followed the pattern above, and was able to merge the new test case i wrote into the existing `_basic` test case with the parametrized fixture. let me know if this is what you had in mind as well